File: MyScript/edk/c/examples/inksearch/InkSearching.c
/**
 * The `InkSearching` example code shows you how to perform ink searching on
 * digital ink that has already been indexed.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptInkSearch.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// path of the file where the index is stored.
#define INDEX_FILE OUTPUT_DIR "/InkIndexing-output.index"

// main function
int main()
{
  voIndex index = NULL;
  voQuery query1 = NULL;
  voQuery query2 = NULL;
  voQuery query = NULL;
  voString word1 = {"Vision", -1};
  voString word2 = {"Objects", -1};
  voFinder finder = NULL;
  voFindResult result = NULL;
  voOccurrenceIterator occurrences = NULL;
  int i = 0;

  if (strlen(INDEX_FILE) == 0)
  {
    printf("/!\\ please edit this example program setting the INDEX_FILE variable to a valid file name\n");
    exit(EXIT_FAILURE);
  }
  
  if (!fileExists(INDEX_FILE))
  {
    printf("/!\\ please run StructuredInputInkIndexing first in order to generate the index file which is needed to run this example\n");
    exit(EXIT_FAILURE);
  }

  // -- library initialization -------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) creating a BinaryQuery (\"Vision OR Objects\")\n");
  query1 = voCreateStringQuery(engine, NULL, &word1, true, true, false);
  verify(query1 != NULL);
  printf(" . \"Vision\" StringQuery created successfully\n");
  query2 = voCreateStringQuery(engine, NULL, &word2, true, true, false);
  verify(query2 != NULL);
  printf(" . \"Objects\" StringQuery created successfully\n");
  query = voCreateBinaryQuery(engine, VO_OrQuery, query1, query2);
  verify(query != NULL);
  printf(" . \"Vision OR Objects\" OrQuery created successfully\n");
  printf("\n");

  printf("3) creating a Finder\n");
  finder = voCreateFinder(engine);
  verify(finder != NULL);
  printf(" . Finder created successfully\n");
  printf("\n");

  printf("4) attaching the OrQuery to the Finder\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, finder, query));
  printf(" . OrQuery attached successfully\n");
  printf("\n");

  printf("5) loading the Index generated from the indexing example\n");
  index = loadEngineObject(INDEX_FILE);
  printf(" . Index loaded successfully\n");
  printf("\n");

  printf("6) attaching the Index to the Finder\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, finder, index));
  printf(" . Index attached successfully\n");
  printf("\n");

  printf("7) running the search process\n");
  verify(VO_IFACE(engine, IRunnable)->run(engine, finder, NULL, NULL));
  printf(" . search process performed successfully\n");
  printf("\n");

  printf("8) retrieving the FindResult\n");
  result = VO_IFACE(engine, IFindResultProvider)->getResult(engine, finder);
  verify(result != NULL);
  printf(" . FindResult retrieved successfully\n");
  printf("\n");

  printf("9) displaying the result\n");
  printf(" . global score: %g\n", VO_IFACE(engine, IFindResult)->getGlobalScore(engine, result));
  occurrences = VO_IFACE(engine, IFindResult)->getOccurrences(engine, result);
  verify(occurrences != NULL);
  for (i = 1; !VO_IFACE(engine, IOccurrenceIterator)->isAtEnd(engine, occurrences); ++i)
  {
    voInputRange inputRange = {NULL, 0};
    verify(VO_IFACE(engine, IOccurrenceIterator)->getInputRange(engine, occurrences, &inputRange));
    inputRange.elements = malloc(sizeof(*inputRange.elements)*inputRange.count);
    verify(VO_IFACE(engine, IOccurrenceIterator)->getInputRange(engine, occurrences, &inputRange));
    printf(" . occurrence %d: score=%g, inputRange=",
      i,
      VO_IFACE(engine, IOccurrenceIterator)->getScore(engine, occurrences));
    printInputRange(inputRange);
    printf(", searchUnit=%d\n",
      (int)VO_IFACE(engine, IOccurrenceIterator)->getSearchUnit(engine, occurrences));
    free(inputRange.elements);
    verify(VO_IFACE(engine, IOccurrenceIterator)->next(engine, occurrences));
  }
  printf("\n");

  printf("10) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, occurrences));
  verify(voReleaseObject(engine, result));
  verify(voReleaseObject(engine, finder));
  verify(voReleaseObject(engine, query));
  verify(voReleaseObject(engine, query2));
  verify(voReleaseObject(engine, query1));
  verify(voReleaseObject(engine, index));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform ink searching\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}

File: MyScript/edk/java/examples/inksearch/InkSearching.java
package inksearch;

import java.io.File;

import certificates.MyCertificate;

import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.inksearch.FindResult;
import com.myscript.inksearch.Finder;
import com.myscript.inksearch.Index;
import com.myscript.inksearch.OccurrenceIterator;
import com.myscript.inksearch.OrQuery;
import com.myscript.inksearch.StringQuery;
import common.CFG;

/**
 * The <code>InkSearching</code> example code shows you how to perform ink
 * searching on digital ink that has already been indexed.
 */
public final class InkSearching
{
  /**
   * The file where the index will be stored.
   */
  private static final String INDEX_FILE = CFG.OUTPUT_DIR + "/InkIndexing-output.index";

  /**
   * Main entry point for the <code>InkSearching</code> example program.
   *
   * @param args command line arguments, ignored.
   */
  public static final void main(String[] args)
  {
    if (INDEX_FILE == null || INDEX_FILE.length() == 0)
    {
      System.err.println("/!\\ please choose a valid filename for index file. You can choose index filename by editing INDEX_FILE static member variable of InkSearching class");
      System.exit(-1);
    }

    if (!(new File(INDEX_FILE)).exists())
    {
      System.err.println("/!\\ please run StructuredInputInkIndexing first in order to generate the index file which is needed to run this example");
      System.exit(-1);
    }

    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) creating a BinaryQuery (\"Vision OR Objects\")");
    final StringQuery query1 = StringQuery.create(engine, "Vision", true, true, false);
    System.out.println(" . \"Vision\" StringQuery created successfully");
    final StringQuery query2 = StringQuery.create(engine, "Objects", true, true, false);
    System.out.println(" . \"Objects\" StringQuery created successfully");
    final OrQuery query = OrQuery.create(engine, query1, query2);
    System.out.println(" . \"Vision OR Objects\" OrQuery created successfully");
    System.out.println();

    System.out.println("3) creating a Finder");
    final Finder finder = Finder.create(engine);
    System.out.println(" . Finder created successfully");
    System.out.println();

    System.out.println("4) attaching the OrQuery to the Finder");
    finder.attach(query);
    System.out.println(" . OrQuery attached successfully");
    System.out.println();

    System.out.println("5) loading the Index generated from indexing example");
    Index index = (Index)EngineObject.load(engine, InkSearching.INDEX_FILE);
    System.out.println(" . Index loaded successfully");
    System.out.println();

    System.out.println("6) attaching the Index to the Finder");
    finder.attach(index);
    System.out.println(" . Index attached successfully");
    System.out.println();

    System.out.println("7) running the search process");    
    finder.run();
    System.out.println(" . search process performed successfully");
    System.out.println();

    System.out.println("8) retrieving the FindResult");
    final FindResult result = finder.getResult();
    System.out.println(" . FindResult retrieved successfully");
    System.out.println();

    System.out.println("9) displaying the result");
    System.out.println(" . global score: " + result.getGlobalScore());
    final OccurrenceIterator occurrences = result.getOccurrences();
    for (int i = 1; !occurrences.isAtEnd(); ++i)
    {
      System.out.println(" . Occurrence " + i +
                         ": score=" + occurrences.getScore() +
                         ", inputRange=" + occurrences.getInputRange() +
                         ", searchUnit=" + occurrences.getSearchUnit() );

      occurrences.next();
    }
    System.out.println();

    // do not forget the 1st rule: you create it, you dispose of it!
    System.out.println("10) cleaning-up: disposing of objects and disposing of the Engine");
    occurrences.dispose();
    finder.dispose();
    query.dispose();
    query2.dispose();
    query1.dispose();
    index.dispose();
    result.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println();

    System.out.println("end of example:");
    System.out.println(" you just learnt how to perform ink searching");
  }

} // InkSearching
